plutus-tx/compiler/Language/PlutusTx/Compiler/Type.hs

```diff
@@ -89,36 +90,33 @@ convTyCon tc = do
             let deps = fmap GHC.getName usedTcs ++ fmap GHC.getName dcs
 
             tvd <- convTcTyVarFresh tc
-            matchName <- safeFreshName () $ (T.pack $ GHC.getOccString $ GHC.getName tc) <> "_match"
 
-            let fakeDatatype = PIR.Datatype () tvd [] matchName []
-            PIR.defineDatatype tcName (PIR.Def tvd fakeDatatype) Set.empty
+            case GHC.unwrapNewTyCon_maybe tc of
+                Just (_, underlying, _) -> do
+                    -- See Note [Coercions and newtypes]
+                    -- Type variables are in scope for the rhs of the alias
+                    alias <- mkIterTyLamScoped (GHC.tyConTyVars tc) $ blackhole (GHC.getName tc) $ convType underlying
+                    PIR.defineType tcName (PIR.Def tvd alias) (Set.fromList deps)
+                    PIR.recordAlias @GHC.Name @() tcName
+                    pure alias
+                Nothing -> do
+                    matchName <- safeFreshName () $ (T.pack $ GHC.getOccString $ GHC.getName tc) <> "_match"
 
-            -- type variables are in scope for the rest of the definition
-            withTyVarsScoped (GHC.tyConTyVars tc) $ \tvs -> do
-                constructors <- for dcs $ \dc -> do
-                    name <- convNameFresh (GHC.getName dc)
-                    ty <- mkConstructorType dc
-                    pure $ PIR.VarDecl () name ty
+                    let fakeDatatype = PIR.Datatype () tvd [] matchName []
```

Comment:
Why do we need to fake a data type here?

Reply:
This is how we handle recursive types - we provisionally insert a datatype that's good enough that we can create a variable reference when we see recursive uses. This deserves a note, I'll add one.